[SPI] Extend SPI interface support in configuration

tonhuisman opened the pull request:
- Added plugin types for 2 and 3 pin SPI configurations (4 pins seems 'illegal' use of Settings.TaskDevicePort field)
- Display SPI configuration and GPIO conflicts with SPI interface in Devices page
- Add exclusions for GPIO pins configured for SPI in standard pin selection
- Added/re-arranged 'missing' stuff to SPI plugins P039 (Thermocouple), P095 (TFT ILI9341) and P096 (eInk)

ESP32 VSPI/HSPI is accounted for.

More improvements & suggestions are welcome!

(I intend to add some SPI plugins in the future)

tonhuisman committed:
[SPI] Extend SPI interface support in configuration

tonhuisman committed:
[SPI] Update SPI-based plugin configuration

TD-er commented:
So, you're quite active with PR's lately :)

tonhuisman commented:
> 
> So, you're quite active with PR's lately :)

It's my holiday, too hot here to do chores around the house (mostly painting), and it's a hobby, so...

tonhuisman committed:
[SPI] Fixed typos in comments...

TD-er commented:
Well, I know the feeling (about the temperature).
Still very much appreciated, all this extra help.

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

TD-er commented on src/WebServer_DevicesPage.ino:
```diff
@@ -395,6 +395,7 @@ void handle_devicess_ShowAllTasksTable(byte page)
     {
       LoadTaskSettings(x);
       struct EventStruct TempEvent;
+      byte spi_gpios[3] = {static_cast<byte>(-1)};
```
This is probably not going to initialize the array to all 0xFF values.
The `= {0};` only works for values of 0.
Also maybe best to use the `int8_t` type instead of `byte`, which is an `uint8_t`.
If you compare to `-1`, I am not entirely sure what will happen here.

tonhuisman replied:
Somewhere in my mind was stored that the GPIOs were all stored as byte in the settings, but I hadn't fact-checked that, and of course that isn't the case. Fixed.
Thanks for the heads-up 👍

TD-er commented on src/WebServer_DevicesPage.ino:
```diff
@@ -565,6 +614,14 @@ void handle_devicess_ShowAllTasksTable(byte page)
+              || spi_gpios[2] == Settings.TaskDevicePin3[x]
+              || Settings.Pin_i2c_sda == Settings.TaskDevicePin3[x]
+              || Settings.Pin_i2c_scl == Settings.TaskDevicePin3[x]) {
+              html += " ";
```
Please use `+= ' ';` instead of `" "` as the last one creates a string in memory compared to a single char.

tonhuisman committed:
[SPI] Minor corrections

tonhuisman committed:
[SPI] Remove unneeded comment

TD-er commented on src/WebServer_DevicesPage.ino:
```diff
@@ -395,7 +395,7 @@ void handle_devicess_ShowAllTasksTable(byte page)
       LoadTaskSettings(x);
       struct EventStruct TempEvent;
-      byte spi_gpios[3] = {static_cast<byte>(-1)};
+      uint8_t spi_gpios[3] {0xFF, 0xFF, 0xFF};  // Keep length at least at 3 or the tests below will fail
```
You're still comparing it to -1, so why not use int8_t?

tonhuisman replied:
My eyes, fingers and brain weren't wired correctly, between reading your earlier review and making the correction.

tonhuisman committed:
[SPI] Minor correction, used wrong type

TD-er committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

tonhuisman committed:
[P096] Widen temp settings var from byte to uint16_t

tonhuisman commented:
The `[P096]` fix is to avoid a crash/strange behavior when configuring resolutions > 255 (as 65535 is allowed in input).

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
[P096] Avoid crash if no eInk display is connected and a yet unhandled command is sent

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman commented:
@TD-er Travis doesn't seem to complete its checks, even though the builds are finished. Is there another way to check that status?

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

tonhuisman committed:
[P095/P096] Allow rotation +270 via cmd mode

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

TD-er commented:
> @TD-er Travis doesn't seem to complete its checks, even though the builds are finished. Is there another way to check that status?

I reported this issue (again) with Travis.
I can "fix" it, but since this is happening every few weeks and it is under investigation with Travis, I guess I should leave it for now.
I do keep an eye on the build logs myself to see what may be failing.

You can also bookmark this link: [Travis builds](https://travis-ci.org/github/letscontrolit/ESPEasy/builds)

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

tonhuisman committed:
[SPI] Correct typo in devices list

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

tonhuisman committed:
[P095] (Re)move touchscreen stuff
There will be a new plugin for XPT2046 touchscreen
Setting wasn't used here anyway

tonhuisman commented:
Started a new plugin for XPT2046, so the setting in P095 is obsolete, and not all those displays have the touchscreen mounted.

@TD-er 
Suggestion for a new plugin number? Now building & testing with P099 (though someone mentioned that number somewhere, but I didn't see a PR yet)
I've already started another Display plugin on P098 (ST7735, similar to P023 OLED plugin), but that's on hold because the hardware didn't arrive yet.

Both P098 and P099 are also waiting for this PR to be merged, as they use/need the improvements implemented here... 😄

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
[P099] Add documentation for Touchscreen
- Device Configuration
- Setting up Calibration
- Configuring Touch objects

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman commented:
Argh, I wanted to commit this to the other PR, I should get some sleep before continuing with this 😞 
Luckily it is only documentation, so no code will break because of this.

tonhuisman committed:
[P099] Remove wrongly committed files

tonhuisman committed:
[P099] Remove link from plugin overview docs

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
[P095] [P096] Avoid using unicode in source

tonhuisman commented:
@TD-er  This PR is (also) complete, can you please review, and merge if suitable, as I have a coupe of dependencies on the improvements I made here. Would be nice to have this finished.

Maybe @jano42 has some thoughts on my adjustments to P095 ILI9341 tft and P096 eInk display?
Other reviews also welcome.

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

jano42 reviewed (commented):
This PR seems to be a nice upgrade effort.
Just reviewing changes (not testing) is good for me (mostly for ILI9341 plugin).
Good job @tonhuisman

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

tonhuisman committed:
Merge branch 'feature/improve_spi_support' of https://github.com/tonhuisman/ESPEasy-1 into feature/improve_spi_support

TD-er commented on docs/source/Controller/C013.rst:
```diff
@@ -247,7 +247,7 @@ These messages are just a serialized byte stream of ``struct C013_SensorInfoStru
     byte header = 255;
     byte ID = 3;
-    byte sourceUnit;
+    byte sourcelUnit;
```
Why change it back?
I renamed it yesterday as I was under the impression it was a typo.

TD-er commented:
I got the feeling you made some error in rebasing it via Git.
This PR now reverts all changes I made the last week.

tonhuisman commented:
I messed up with the commit after merging mega into my branch. Will fix today, sorry for the confusion.

TD-er commented:
If you're using a tool like GitKraken, you can quite easily revert to a previous commit.
Maybe it is best to first checkout the commit that was still OK and create a new branch at that point.
To prepare for the worst, you can still create a new pull request from that branch (so you do have a backup on Github after pushing that new branch) and then try to reset the HEAD of your branch to the right commit.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/improve_spi_support

tonhuisman commented:
> If you're using a tool like GitKraken, you can quite easily revert to a previous commit.
> Maybe it is best to first checkout the commit that was still OK and create a new branch at that point.

This time, GitKraken saved the day, eventually a git reset --hard to the previous state, push to origin, pull upstream mega and apply a couple of manual merges, then commit and push again, seems the way to resolve this. Learned a lot today about git, thanks (hope I don't mess up like this again 😓)

I think it is solved now.

TD-er commented:
> hope I don't mess up like this again 😓

Well if all goes without errors, then it is a sad day as you didn't learn anything.

But it does indeed look OK now, just 7 files in the PR.

tonhuisman commented:
PR is still open for review, as it would be nice to have this merged so I can work on some other SPI plugins.

tonhuisman committed:
Merge branch 'mega' into feature/improve_spi_support

TD-er commented:
I am now looking into this PR, and I want to cleanup the WebServer_DevicesPage.ino file a bit.
Those handle_... functions have become extremely long, so I want to split them into separate functions to make them a bit more readable.
After that I will merge it.

TD-er committed:
[Web Interface] Improve task config consistency + code readability

Split those extremely large functions into separate sub functions to improve readability.
Made GPIO formatting more consistent.

TD-er committed:
[P039 Thermocouple] Add FIXME to remove unneeded strings.
